dbsc/workflow/rules/mrtpipelines.smk

```diff
@@ -1,571 +1,600 @@
-            datatype='dwi',
-            desc='sfwm',
-            suffix='response.txt',
-        )
-        avg_gm=bids(
-            root=join(config['output_dir'], 'mrtpipelines/avg'),
-            datatype='dwi',
-            desc='gm',
-            suffix='response.txt',
-        )
-        avg_csf=bids(
-            root=join(config['output_dir'], 'mrtpipelines/avg'),
-            datatype='dwi',
-            desc='csf',
-            suffix='response.txt',
-        )
-    group: "group"
+        wm_avg_rf=bids(
+            root=join(mrtrix_dir, "avg"),
+            datatype="response",
+            desc="wm",
+            suffix="response.txt",
+        ),
+        gm_avg_rf=bids(
+            root=join(mrtrix_dir, "avg"),
+            datatype="response",
+            desc="gm",
+            suffix="response.txt",
+        ),
+        csf_avg_rf=bids(
+            root=join(mrtrix_dir, "avg"),
+            datatype="response",
+            desc="csf",
+            suffix="response.txt",
+        ),
+    threads: workflow.cores
+    group:
+        "group"
     container:
-        config['singularity']['mrtpipelines']
+        config["singularity"]["mrtrix"]
     shell:
-        'average_response {input.sfwm} {output.avg_sfwm} {input.gm} {output.avg_gm} {input.csf} {output.csf}'
+        "responsemean {input.wm_rf} {output.wm_avg_rg} -nthreads {threads} &&"
+        "responsemean {input.gm_rf} {output.gm_avg_rg} -nthreads {threads} &&"
+        "responsemean {input.csf_rf} {output.csf_avg_rg} -nthreads {threads}"
 
 
-rule compute_fod:
+rule dwi2fod:
+    """Jeurissen, B; Tournier, J-D; Dhollander, T; Connelly, A & Sijbers, J. Multi-tissue constrained spherical deconvolution for improved analysis of multi-shell diffusion MRI data. NeuroImage, 2014, 103, 411-426"""
     input:
-        avg_sfwm=rules.avg_response.output.avg_sfwm,
-        avg_gm=rules.avg_response.output.avg_gm,
-        avg_csf=rules.avg_response.output.avg_csf,
-        dwi=rules.nii_to_mif.output.dwi,
-        mask=rules.nii_to_mif.output.mask,
+        dwi=rules.nii2mif.output.dwi,
+        mask=rules.nii2mif.output.mask,
+        wm_rf=join(config["responsemean_dir"], "desc-wm_response.txt")
+        if responsemean_flag
+        else rules.responsemean.output.wm_avg_rf,
+        gm_rf=join(config["responsemean_dir"], "desc-gm_response.txt")
+        if responsemean_flag
+        else rules.responsemean.output.gm_avg_rf,
+        csf_rf=join(config["responsemean_dir"], "desc-csf_response.txt")
+        if responsemean_flag
+        else rules.responsemean.output.csf_avg_rf,
     params:
-        shell=config.get('shells', ''),
-        lmax=config.get('lmax', ''),
-        threads=workflow.threads,
+        shell=shells,
     output:
         wm_fod=bids(
-            root=join(config['output_dir'], 'mrtpipelines'),
-            datatype='fod',
-            model='csd',
-            desc='wm',
-            suffix='fod.mif'
+            root=mrtrix_dir,
```

Comment:
It would save a bit of space to define a variable outside the rule with the `desc` as a wildcard, then expand over it for each output. Something like:

```
dwi2fod_out = bids(
    root=mrtrix_dir,
    datatype="response",
    model="csd",
    desc="{desc}",
    suffix="fod.mif",
    **config["subj_wildcards"],
)

rule dwi2fod:
    ...
    output:
        wm_fod=expand(dwi2fod_out, desc="wm", allow_missing=True),
    ....
```

Reply:
This applies for a few of the following rules, too.

Reply:
Would that still work to define it as a wildcard if the different tissue types still need to be explicitly defined in the script (in order of `{input} {output} {input} {output} ...`)

Reply:
Yeah, to be clear I would still define each of `wm_fod`, `gm_fod`, and `csf_fod` as inputs, I would just define them this way instead of repeating the (almost identical) `bids` definition every time

Reply:
Great, I went ahead and updates to do something similar to this for most of the pre-processing rules. Not sure if its the best idea yet for the tractography rules.